src/rustllvm/RustWrapper.cpp

```diff
@@ -958,3 +958,27 @@ LLVMRustBuildLandingPad(LLVMBuilderRef Builder,
     return LLVMBuildLandingPad(Builder, Ty, PersFn, NumClauses, Name);
 #endif
 }
+
+extern "C" void
+LLVMRustHookUnwindResume(LLVMModuleRef Mod) {
+    Module *M = unwrap(Mod);
+    LLVMContext &Ctx = M->getContext();
+
+    FunctionType *FTy = FunctionType::get(Type::getVoidTy(Ctx), Type::getInt8PtrTy(Ctx), false);
+    Constant *RustRewind = M->getOrInsertFunction("rust_eh_unwind_resume", FTy);
+
+    Function *OrigRewind = Function::Create(FTy, GlobalVariable::InternalLinkage, "_Unwind_Resume", M);
+    BasicBlock *BB = BasicBlock::Create(Ctx, "", OrigRewind);
+    CallInst *CI = CallInst::Create(RustRewind, &*OrigRewind->arg_begin(), "", BB);
+    CI->setCallingConv(CallingConv::C);
+    // We never expect rust_eh_unwind_resume to return.
+    new UnreachableInst(Ctx, BB);
+
+    // Until DwarfEHPrepare pass has run, _Unwind_Resume is not referenced by any live code
+    // and is subject to dead code elimination.  Here we add _Unwind_Resume to @llvm.globals to prevent that.
+    PointerType *elem_type = PointerType::get(IntegerType::get(Ctx, 8), 0);
+    ArrayType *used_type = ArrayType::get(elem_type, 1);
+    Constant* init = ConstantArray::get(used_type, ArrayRef<Constant*>(ConstantExpr::getBitCast(OrigRewind, elem_type)));
+    GlobalVariable* used = new GlobalVariable(*M, used_type, false, GlobalValue::AppendingLinkage, init, "llvm.used");
+    used->setSection("llvm.metadata");
+}
```

Comment:
Can this all be written in Rust? It's generally quite difficult to keep C++ up to date across LLVM versions and the C API is much more stable.